### ISSUE TITLE
Filtering out zero/null values in tooltip and sort the tooltip by field values 

### ISSUE THREAD

**Issue by sfc-gh-cheliu** [vega/vega-lite#9152]
I'd like to add a feature to **hide/filter out zero/null values** in the tooltip, and potentially **sort the tooltip by the value of each field/column**. Here's an example:
<img width="401" alt="example" src="https://github.com/vega/vega-lite/assets/112987415/3a9f8650-5616-417c-9f4f-a613705d37d2">
Basically I want to hide `type3` and `type5` in the tooltip, and sort the remaining by descending order of Value.

Since I'm not familiar with altair/vega libraries, so this might not be the best approach, but my thought is to add a new parameter/condition in altair and passing this to the tooltip object, so that when the frontend receives this signal, it will hide the values I don't want to see in the tooltip and possibly sort the values.

I found 2 places that can be modified:
1. https://github.com/vega/vega-lite/blob/4ba3b93e575df97ffe56713287b40d6c55b18340/src/compile/mark/encode/tooltip.ts
2. https://github.com/vega/vega-tooltip/blob/5f805fb8a6e76e3e58d2541e51b6106bcc133f11/src/Handler.ts

Any guidance and suggestions are welcome. Thanks! cc @joelostblom @jakevdp as well as core vega-lite developers @kanitw @domoritz @arvind @jheer

**Comment by sfc-gh-cheliu** [comment]
@domoritz I agree. I've pushed some new commits (making 'sort' to be a new parameter as part of 'encoding') in https://github.com/vega/vega-lite/pull/9148 with more details in the description. Appreciate it if you could take a look.

Yeah sorting is trickier. From my understanding, before the encoding defs are passed to vega-tooltip they remain to be some vega expressions, and their values cannot be determined. I really spent some time investigating whether there's a intermediate step to convert these expressions to actual values. But it seems that the values will only be evaluated when the mouse hover event is triggered, that's why I modify the things in vega-tooltip as well. Correct me if I'm wrong, and if you have a way to handle everything within vega-lite & vega, that'd be great:)

**Comment by domoritz** [comment]
Note that Vega tooltip is only one example implementation of a tooltip for Vega. Therefore I think we should try to do as much as possible in Vega/Vega-Lite. As you develop your code, test it without Vega tooltip to see what the default tooltips (using HTML `title`) look like for example.

Can you summarize the proposed API changes here? It'll be easier to discuss it compared to the pull request. I know it's a bit more work but I would appreciate it. Thanks!

**Comment by sfc-gh-cheliu** [comment]
@domoritz Thanks for the reply. 
1) QQ: is specifying multiple fields only supported in vega-tooltip? I vaguely remember seeing an issue asking for this feature and I suppose before vega-tooltip is out the default tooltip can only contain one field. If this is the case, I guess we can only add this feature to vega-tooltip because it doesn't really make a lot of sense to sort a single field.

2) Sure, I summarize it in the description of https://github.com/vega/vega-lite/pull/9148. Copying it here for your reference:
Parameter signatures:
Filter - added as a parameter for each tooltip object in the tooltip array:
```
{"field": "type1", "type": "quantitative", "**filter**": {"operator": ">", "literal": 0}
```

Sort - added as a parameter for encoding:
```
"encoding": { "opacity": { "condition": { "empty": false, "param": "param_46", "value": 0.2 }, "value": 0 }, "size": { "value": 4 }, "tooltip": [ { "field": "Date", "type": "nominal" }, { "field": "type1", "type": "quantitative" }, { "field": "type2", "type": "quantitative" }, { "field": "type3", "type": "quantitative" }, { "field": "type4", "type": "quantitative" }, { "field": "type5", "type": "quantitative" }, { "field": "type6", "type": "quantitative" }, { "field": "type7", "type": "quantitative" }, { "field": "type8", "type": "quantitative" } ], "**sort_tooltip**": {"value": "descending"}, "x": { "field": "Date", "title": null, "type": "ordinal" } }, "mark": { "type": "rule" }, "params": [ { "name": "param_46", "select": { "clear": "mouseout", "fields": [ "Date" ], "nearest": true, "on": "mouseover", "type": "point" } } ] 
```

To make the new parameters robust, both filter & sort_tooltip can only be some enum values.
filter.operator: one of '==', '!=', '<', '<=', '>', '>='
filter.literal: one of type string, number or boolean.
```"filter": {"operator": "><", "literal": 0}:```
![image](https://github.com/vega/vega-lite/assets/112987415/03ea799a-1873-4b99-a39a-f3055dc768b6)


sort_tooltip.value: 'ascending' or 'descending' only:
```"sort_tooltip": {"value": "whatever"}```
![image](https://github.com/vega/vega-lite/assets/112987415/bd61fe90-976e-43b1-a536-afc526183b62)

Please see https://github.com/vega/vega-lite/pull/9148 for more details. Thanks!

**Comment by kanitw** [comment]
<img width="684" alt="image" src="https://github.com/vega/vega-lite/assets/111269/dc17399b-6cc6-4913-af74-b4e841ba1935">

I think this reads more like the spec is filtering the field in general, rather than in tooltip?

**Comment by sfc-gh-cheliu** [comment]
@kanitw Thanks for the reply! 
I read through the vega-lite implementation and find that tooltip object just implements some higher level interface/type, so they might have overlap with other fields. However, only the fields specified within the 'tooltip' field can be applied with a filter. For instance, if I add the filter to the 'size' field, it will throw an warning:
`"size": { "value": 4, "filter": {"operator": ">", "literal": 0}}`
<img width="1666" alt="Screenshot 2023-10-28 at 12 00 22 PM" src="https://github.com/vega/vega-lite/assets/112987415/569d0617-bb20-4448-b4f4-15712ea469d3">

**Comment by sfc-gh-cheliu** [comment]
@domoritz Made a PR to support default tooltips: https://github.com/vega/vega/pull/3816. Let me know if there's anything else we should do. Thanks!

**Comment by sfc-gh-cheliu** [comment]
@domoritz @kanitw I hope you are doing well! Could we push this forward a little bit? Lemme know anything that you think can be done/improved from my side. Thanks again!

**Comment by sfc-gh-cheliu** [comment]
@domoritz @kanitw Any update on this? I want to roll this out in the near future. Please lemme know if there's anything remaining to be handled. Thanks!

**Comment by sfc-gh-cheliu** [comment]
@domoritz @kanitw Could we unblock this? Thanks.

**Comment by sfc-gh-cheliu** [comment]
@domoritz @kanitw Kindly follow up on this again. Thanks.

**Comment by sfc-gh-cheliu** [comment]
@domoritz @kanitw Happy new year folks! I hope you have a wonderful break. Could we move this forward a little bit? I believe this will be a useful feature for the whole community since I've seen such requests even as early as 3 years ago. So, it'd be good if you could review this so that I can make further changes. Thanks!

**Comment by sfc-gh-cheliu** [comment]
@domoritz @kanitw Could we proceed with this? This feature will be super helpful for my team and the whole altair community if we could roll it out. Thanks.

**Comment by domoritz** [comment]
I'll put it on my queue. Thanks for the ping.

**Comment by sfc-gh-cheliu** [comment]
@domoritz Hi Dominik, I hope you are doing well. Kindly check in to see if you get any chance to take a look at this. Thanks!

**Comment by domoritz** [comment]
Uhh, yeah, it's in my queue. I'm going on vacation tomorrow for a few weeks but let me bump it up for after I return.

**Comment by sfc-gh-cheliu** [comment]
@domoritz Appreciate that, have a wonderful vacation!

**Comment by domoritz** [comment]
#9876 addresses filtering. If sorting etc is needed, let's make a new issue.
